File: ios/Sources/NetUtilsPlugin/NetUtilsPlugin.swift
import Foundation
import Capacitor

@objc(NetUtilsPlugin)
public class NetUtilsPlugin: CAPPlugin, CAPBridgedPlugin {

  public let identifier = "NetUtilsPlugin"
  public let jsName = "NetUtils"

  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "getInterfaces", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "checkUrl", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "checkPort", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "resolveHostname", returnType: CAPPluginReturnPromise),

    CAPPluginMethod(name: "startForwarding", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "stopForwarding", returnType: CAPPluginReturnPromise),

    CAPPluginMethod(name: "sshExecSync", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sshConnect", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sshStartShell", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sshWrite", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sshDisconnect", returnType: CAPPluginReturnPromise),

    CAPPluginMethod(name: "tcpConnect", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "tcpWrite", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "tcpDisconnect", returnType: CAPPluginReturnPromise),
  ]

  private let netUtils = NetUtils()
  private let sshUtils = SSHUtils()
  private let portForwarder = PortForwarder()
  private var sshSession: SSHSession?
  private var tcpSocket: TCPSocket?

  // basic

  @objc func getInterfaces(_ call: CAPPluginCall) {
    netUtils.getInterfaces(call)
  }

  @objc func checkUrl(_ call: CAPPluginCall) {
    netUtils.checkUrl(call)
  }

  @objc func checkPort(_ call: CAPPluginCall) {
    netUtils.checkPort(call)
  }

  @objc func resolveHostname(_ call: CAPPluginCall) {
    netUtils.resolveHostname(call)
  }

  @objc func startForwarding(_ call: CAPPluginCall) {
    portForwarder.startForwarding(call)
  }

  @objc func stopForwarding(_ call: CAPPluginCall) {
    portForwarder.stopForwarding(call)
  }

  // ssh

  @objc func sshExecSync(_ call: CAPPluginCall) {
    sshUtils.execSync(call)
  }

  @objc func sshConnect(_ call: CAPPluginCall) {
    sshSession?.cleanup()
    sshSession = SSHSession(plugin: self)
    sshSession!.connect(call)
  }

  @objc func sshStartShell(_ call: CAPPluginCall) {
    guard let sshSession = sshSession else {
      call.reject("No active connection to start shell on.")
      return
    }

    sshSession.startShell(call)
  }

  @objc func sshWrite(_ call: CAPPluginCall) {
    guard let sshSession = sshSession else {
      call.reject("No active connection to write to.")
      return
    }

    sshSession.write(call)
  }

  @objc func sshDisconnect(_ call: CAPPluginCall) {
    guard let sshSession = sshSession else {
      call.reject("No active connection to disconnect.")
      return
    }

    sshSession.disconnect(call)
    self.sshSession = nil
  }

  // tcp

  @objc func tcpConnect(_ call: CAPPluginCall) {
    tcpSocket?.disconnect(nil)
    tcpSocket = TCPSocket(plugin: self)
    tcpSocket!.connect(call)
  }

  @objc func tcpWrite(_ call: CAPPluginCall) {
    guard let tcpSocket = tcpSocket else {
      call.reject("No active connection to write to.")
      return
    }

    tcpSocket.write(call)
  }

  @objc func tcpDisconnect(_ call: CAPPluginCall) {
    guard let tcpSocket = tcpSocket else {
      call.reject("No active connection to disconnect.")
      return
    }

    tcpSocket.disconnect(call)
    self.tcpSocket = nil
  }
}

File: android/src/main/java/com/macchie/plugins/netutils/NetUtilsPlugin.java
package com.macchie.plugins.netutils;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "NetUtils")
public class NetUtilsPlugin extends Plugin {

  private final NetUtils _netUtils = new NetUtils();
  private final SSHUtils _sshUtils = new SSHUtils();
  private final PortForwarder _portForwarder = new PortForwarder();
  private SSHSession _sshSession;
  private TCPSocket _tcpSocket;

  public void emit(String eventName, JSObject data) {
    notifyListeners(eventName, data);
  }

  // basic

  @PluginMethod
  public void getInterfaces(PluginCall call) {
    _netUtils.getInterfaces(call);
  }

  @PluginMethod
  public void checkUrl(PluginCall call) {
    _netUtils.checkUrl(call);
  }

  @PluginMethod
  public void checkPort(PluginCall call) {
    _netUtils.checkPort(call);
  }

  @PluginMethod
  public void resolveHostname(PluginCall call) {
    _netUtils.resolveHostname(call);
  }

  // forwarding

  @PluginMethod
  public void startForwarding(PluginCall call) {
    _portForwarder.startForwarding(call);
  }

  @PluginMethod
  public void stopForwarding(PluginCall call) {
    _portForwarder.stopForwarding(call);
  }

  // ssh

  @PluginMethod
  public void sshExecSync(PluginCall call) {
    _sshUtils.sshExecSync(call);
  }

  @PluginMethod
  public void sshConnect(PluginCall call) {
    _sshSession = new SSHSession(this);
    _sshSession.connect(call);
  }

  @PluginMethod
  public void sshStartShell(PluginCall call) {
    if (_sshSession == null) {
      call.reject("No active connection to start shell on.");
      return;
    }
    _sshSession.startShell(call);
  }

  @PluginMethod
  public void sshWrite(PluginCall call) {
    if (_sshSession == null) {
      call.reject("No active connection to write to.");
      return;
    }
    _sshSession.write(call);
  }

  @PluginMethod
  public void sshDisconnect(PluginCall call) {
    if (_sshSession == null) {
      call.reject("No active connection to disconnect.");
      return;
    }
    _sshSession.disconnect(call);
  }

  // tcp

  @PluginMethod
  public void tcpConnect(PluginCall call) {
    _tcpSocket = new TCPSocket(this);
    _tcpSocket.connect(call);
  }

  @PluginMethod
  public void tcpWrite(PluginCall call) {
    if (_tcpSocket == null) {
      call.reject("No active connection to write to.");
      return;
    }
    _tcpSocket.write(call);
  }

  @PluginMethod
  public void tcpDisconnect(PluginCall call) {
    if (_tcpSocket == null) {
      call.reject("No active connection to disconnect.");
      return;
    }
    _tcpSocket.disconnect(call);
  }
}
